rfd/0133-connect-my-computer.md

```diff
@@ -0,0 +1,513 @@
+function to avoid the downloading logic being spread across files and listeners.
+
+We have to be mindful about keeping the preload APIs as constrained as possible. The renderer should
+not get the ability to download an arbitrary file to an arbitrary path. If possible, the download
+process should be totally encapsulated from the renderer, with the renderer merely signaling the
+intent to start the agent and the main process handling everything else.
+
+##### Storing the node agent
+
+The binary is stored in `~/Library/Caches/Teleport Connect` on macOS and `~/.cache/Teleport Connect`
+on Linux. It is extracted to a directory of the same name as the downloaded tar.gz archive, for
+example `teleport-v13.0.4-darwin-arm64-bin`. As the parent directory can be cleared by the OS or the
+user, before each launch Connect is going to check if the agent exists and download it automatically
+if not.
+
+So far we’ve been storing all data in Electron’s `userData` which is `~/Library/Application Support`
+on macOS. [That directory is not suited for storing large files](https://www.electronjs.org/docs/latest/api/app#appgetpathname) as it might be backed up to
+cloud storage. 
+
+On top of that, the agent binary is something that can be easily recreated if needed by downloading
+it again. After reading Apple Developer docs ([Optimizing Your App’s Data for iCloud Backup](https://developer.apple.com/documentation/foundation/optimizing_your_app_s_data_for_icloud_backup/#3928528)
+and archived [File System Programming Guide](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW4)),
+we’ve decided that `~/Library/Caches` is a suitable directory for storing the agent.
+
+Out of the box, Electron doesn’t expose `~/Library/Caches` through its `app.getPath` function,
+unlike `~/Library/Application Support` which is retrievable by calling `app.getPath('userData')`.
+
+A good candidate for that would be `sessionData` which the docs describe as “the directory for
+storing data generated by Session, such as localStorage, cookies, disk cache, downloaded
+dictionaries, network state, devtools files.” Unfortunately, for legacy reasons `sessionData` points
+to the same directory as `userData`. [The docs recommend changing the location](https://www.electronjs.org/docs/latest/api/app#appgetpathname) of this directory to
+another location.
+
+There’s [an outstanding issue about Electron storing its caches in the wrong location](https://github.com/electron/electron/issues/8124). Based on the Apple Developer docs and
+the discussion under that issue, changing the location of `sessionData` to `~/Library/Caches` on
+macOS and `~/.cache` on Linux seems like a reasonable thing to do. As we have not been using cookies
+or localStorage thus far it should be safe to do so with no risk of losing any user-generated state.
+
```

Comment:
We wanted Connect to manage the lifecycle of the agent. The way I think about this feature is that since it allows access to _your_ device as _your_ system user, the agent should not run "unattended". The moment you close Connect, the agent should also stop.

In a situation where you do "Restart the system and reopen windows when logging back in", if the OS launches Connect the agent will boot up automatically with Connect (though I haven't tested this scenario yet).

---

Matter of fact, one of the things we still don't know how to do is how to kill the agent process if the Electron app is killed. For the MVP, we considered ignoring this problem and investigating what happens in that situation if you try to relaunch Connect.

Reply:
As for the problem of killing the agent if the Electron app dies, I found the following:

Since we can modify the agent (the teleport binary), we could pass the pid of the Electron app to the agent and start a goroutine in the agent which would periodically check if the parent pid has changed. [A similar approach has been described on SO](https://stackoverflow.com/a/23587108/742872) in the second half of the answer. I'm not sure if this is something we should do for the MVP though.

As for the behavior of the agent when a zombie agent is left running:

If another agent is started that uses the same data dir and the config as one that is already running, any new connections will be forwarded to the new agent. Once the agent is stopped, the old agent starts picking connections back again.

Removing the data dir and the config of a running agent does not cause the agent to detect an anomaly and shut down, but it does prevent it from accepting new connections.

Of course, I'd consider this behavior to be undefined since it's not behavior that other parts of the system depend on.

Reply:
> since it allows access to your device as your system user, the agent should not run "unattended". The moment you close Connect, the agent should also stop.

I love this logic. Can we mention this explicitly in the RFD?

Reply:
> I love this logic. Can we mention this explicitly in the RFD?

Yeah, I already kind of added it to [the Security section](https://github.com/gravitational/teleport/blob/rfd/0133-connect-my-computer/rfd/0133-connect-my-computer.md#access-to-ui-and-autostart):

> (…) if the agent is running, the user must be able to stop the agent without being forced to close the app. (…) if the user can't access the UI to control the agent, the agent should not be running.

But I should expand it with why we don't use systemd to run the agent unattended.